### ISSUE TITLE
Buildah task would be faster with `--no-cache` than `--layers`

### ISSUE THREAD

**Issue by holly-cummins** [tektoncd/catalog#265]
# Expected Behavior

The buildah task should have the optimum arguments to go fast in a CI/CD environment. 

# Actual Behavior

I can improve buildah build times from ~10m to ~6m by adjusting the cache settings. I don't think investing in caching onto an ephemeral volume is likely to be the right behaviour for most scenarios; even with my multi-stage build performance is faster without the laters.

# Steps to Reproduce the Problem

1. Run a build in a Tekton pipeline, with the builah task. I'm using OpenShift 4.
2. Measure the image build time.
3. Fork the buildah task, replace `--layers` with `--no-cache`, and update the pipeline to use the fork. 
4. Run a build and measure the image build time. 

For my cluster and dockerfile, the update seems to have improved image build times from ~10m to ~6m. 

I don't believe there's much value in the `--layers` option in the task as written, because it uses an `emptyDir()` volume, so any cached layers are ephemeral. Because there's no caching between builds, the cost of writing the layers outweighs the saving from the cache. (https://github.com/containers/buildah/issues/1455 has some discussion of caching in CI/CD environments.)

# Additional Info

I explored the possibility of replacing the `emptyDir{}` volume with a persistent volume claim, but (after a bit of a struggle with storage types and storage drivers) I discovered https://github.com/containers/buildah/issues/2215. I think that means that if if there were a persistent volume, tekton wouldn't use the cache much because the timestamps of the source files are always new. 

This is my Dockerfile, but I don't think there's anything distinctive in it; I think the perf characteristics would be similar no matter what the dockerfile.

```
# base image
FROM node:alpine AS builder
# set working directory
WORKDIR /app
# install client node packages
COPY ./client/package.json /client/yarn.lock ./
RUN yarn install --network-timeout 100000
# yarn build for client
RUN printenv
COPY ./client .
# yarn assumes build is within root directory
RUN yarn --cwd ./client run build
# build our server
FROM node:alpine
WORKDIR /app
# install server node packages
COPY ./server/package.json /server/yarn.lock ./
RUN yarn install --network-timeout 100000
# import build files from client
COPY --from=builder /app/build ./public
RUN printenv
COPY ./server .

#This isn't necessary in a privileged context, or when built with docker, but is needed with buildah
ENV HOME /app

CMD ["yarn", "start"]
```

**Comment by holly-cummins** [comment]
I'm happy to put in a PR for this if there's not a good reason to keep using `--layers`.

**Comment by vdemeester** [comment]
@holly-cummins I think we should make the `varlibcontainer` volume a workspace, allowing the users to provide a PVC or something else in order to make it more useful. We could also have an option to disable cache (and thus not using `--layers`)

**Comment by holly-cummins** [comment]
Using a workspace was my first thought as well. . I did experiment with providing a PVC in the varlibcontainer, but there were two problems.
* Buildah played very badly with most PVs. I don't have enough storage expertise to debug fully, but what I found was that file storage PVs were so slow that a 10m build would take 5-10 hours. Block storage behaved better, but the lack of parallelism is no good in a CI/CD context. It seems like buildah is quite sensitive to the filesystem. I ended up hitting https://github.com/containers/buildah/issues/1481 and doing a lot of experimentation with `--storage-driver` options, before giving up, because I realised ... 
* Buildah layer caching is near-useless in a CI/CD environment (https://github.com/containers/buildah/issues/2215). For COPY, buildah looks at the file timestamps to work out what changed. Locally, that works fine, in a CI/CD system, with a fresh git clone, every file will be reported as changed. In my builds, I got one layer (a `WORKDIR`) cached, and then the cache wasn't used. 

So long term, once https://github.com/containers/buildah/issues/2215 is fixed, using a workspace would be helpful, but unless I've misunderstood the cache issues, I don't see value short term. 

I agree that an option makes sense, since there may be some use cases where the caching does help. I'd vote for caching disabled by default, because I can't think of a use case, given its limitations, but I'm happy to  be told I missed a use case. :)
